### ISSUE TITLE
value is incorrectly marked as 'Never' in a generator

### ISSUE THREAD

**Issue by decorator-factory** [microsoft/pyright#2225]
### Minimal example:
```py
Callback = Callable[[], None]

def foo():
    answer: Optional[int] = None

    def set_answer():
        nonlocal answer
        answer = 42

    yield set_answer

    assert answer is not None
    reveal_type(answer)
    return answer
```

### Expected behaviour:
- `reveal_type` shows `int` 
- return type of `foo` is inferred as `Generator[() -> None, None, int]` 

### Actual behaviour:
- `reveal_type` shows `Never`
- return type of `foo` is inferred as `Generator[() -> None, None, Unknown]` 

Because I'm yielding a function that can change the value of `answer`, we don't know if `answer` is going to be `None` when control is yielded back.

**Comment by decorator-factory** [comment]
After further investigation, the issue isn't really with generators. In this example: ```py
Callback = Callable[[], None]

def bar(fn: Callback):
    ...

def foo():
    answer: Optional[int] = None

    def set_answer():
        nonlocal answer
        answer = 42

    bar(set_answer)

    assert answer is not None
    return answer
```
`answer` is also inferred as `Never` in the end.

**Comment by decorator-factory** [comment]
I think this is a duplicate of #851
